sbtc-core/src/operations/commit_reveal/utils.rs

```diff
@@ -0,0 +1,171 @@
+/*!
+Utils for commit reveal transaction construction
+*/
+use std::{iter::once, num::TryFromIntError};
+
+use bdk::bitcoin::{
+    blockdata::{
+        opcodes::all::{OP_CHECKSIG, OP_DROP, OP_RETURN},
+        script::Builder,
+    },
+    schnorr::UntweakedPublicKey,
+    secp256k1::Secp256k1,
+    util::taproot::{LeafVersion, TaprootBuilder, TaprootBuilderError, TaprootSpendInfo},
+    Address as BitcoinAddress, Network, OutPoint, PackedLockTime, Script, Sequence, Transaction,
+    TxIn, TxOut, Witness, XOnlyPublicKey,
+};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+/// Commit reveal error
+pub enum CommitRevealError {
+    #[error("Signature is invalid: {0}")]
+    /// Invalid recovery ID
+    InvalidRecoveryId(TryFromIntError),
+    #[error("Control block could not be built from script")]
+    /// No control block
+    NoControlBlock,
+    #[error("Could not build taproot spend info: {0}: {1}")]
+    /// Taproot error
+    InvalidTaproot(&'static str, TaprootBuilderError),
+}
+
+/// Commit reveal result
+pub type CommitRevealResult<T> = Result<T, CommitRevealError>;
+
+fn internal_key() -> UntweakedPublicKey {
+    // Copied from BIP-0341 at https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs
+    // The BIP recommends a point lift_x(0x0250929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0).
+    // This hex string is copied from the lift_x argument with the first byte stripped.
+
+    // TODO: Verify that this point is secure
+    let internal_key_vec =
+        hex::decode("50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0").unwrap();
+
+    // TODO: Confirm that this is infallible
```

Comment:
Since we have unit tests this is infallible
```suggestion
```

Reply:
Oh actually we don't. Should we perhaps add some unit tests?

Reply:
Perhaps a big ask, but ideally we'd have a unit test spinning up a bitcoin regtest node, constructing and broadcasting a commit UTXO, and constructing a reveal for it.

Reply:
Discussed the testing plan on Slack:

- We test parsing of transactions with some custom tx output generation code (similar to what I did for op return parsing module tests)
- We then write real construction logic for the same tx types and replace the custom tx output generation code - then use parsing logic to test these
- In some way construction logic would test the parsing logic, and vice versa
- We could then extend the testing situation by spinning a BTC node and trying to submit these transactions